src/sentry/integrations/source_code_management/repository.py

```diff
@@ -1,39 +1,98 @@
+from sentry.integrations.base import IntegrationInstallation
 from sentry.integrations.services.repository import RpcRepository
 from sentry.models.identity import Identity
 from sentry.models.repository import Repository
+from sentry.shared_integrations.client.base import BaseApiResponseX
 from sentry.shared_integrations.exceptions import ApiError, IntegrationError
+from sentry.utils import metrics
 
+REPOSITORY_INTEGRATION_CHECK_FILE_METRIC = "repository_integration.check_file.{result}"
+REPOSITORY_INTEGRATION_GET_FILE_METRIC = "repository_integration.get_file.{result}"
 
-class RepositoryMixin:
-    # whether or not integration has the ability to search through Repositories
-    # dynamically given a search query
-    repo_search = False
 
+class RepositoryIntegration(IntegrationInstallation, ABC):
+    @property
+    def codeowners_locations(self) -> list[str] | None:
+        """
+        A list of possible locations for the CODEOWNERS file.
+        """
+        return None
+
+    @property
+    def repo_search(self) -> bool:
+        return True
+
+    @property
+    @abstractmethod
+    def integration_name(self) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_client(self) -> RepositoryClient:
+        """Returns the client for the integration. The client must be a subclass of RepositoryClient."""
+        raise NotImplementedError
+
+    @abstractmethod
     def source_url_matches(self, url: str) -> bool:
-        """Checks if the url matches the integration's source url."""
+        """Checks if the url matches the integration's source url. Used for stacktrace linking."""
         raise NotImplementedError
 
-    def format_source_url(self, repo: Repository, filepath: str, branch: str) -> str:
-        """Formats the source code url used for stack trace linking."""
+    @abstractmethod
+    def format_source_url(self, repo: Repository, filepath: str, branch: str | None) -> str:
+        """Formats the source code url used for stacktrace linking."""
         raise NotImplementedError
 
+    @abstractmethod
     def extract_branch_from_source_url(self, repo: Repository, url: str) -> str:
-        """Extracts the branch from the source code url."""
+        """Extracts the branch from the source code url. Used for stacktrace linking."""
         raise NotImplementedError
 
+    @abstractmethod
     def extract_source_path_from_source_url(self, repo: Repository, url: str) -> str:
-        """Extracts the source path from the source code url."""
+        """Extracts the source path from the source code url. Used for stacktrace linking."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def has_repo_access(self, repo: RpcRepository) -> bool:
+        """Used for migrating repositories. Checks if the installation has access to the repository."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_repositories(self, query: str | None = None) -> Sequence[dict[str, Any]]:
```

Comment:
Possible future improvement: Could we define a proper type to encapsulate an Integration Repository in the future? Having `Any` for this isn't great :/

Reply:
after a quick glance, it seems possible to define a typed dict with some optional elements, but i'd have to look closer